pkg/kube/secrets/manifest.go

```diff
@@ -0,0 +1,241 @@
+		deploymentName,
+	}
+}
+
+func (p *manifestPersisterImpl) PersistManifest(manifest manifest.Manifest, sourceDescription string) (*corev1.Secret, error) {
+	currentVersion, err := getGreatestVersion(p)
+	if err != nil {
+		return nil, err
+	}
+
+	version := currentVersion + 1
+
+	secretName, err := secretName(p, version)
+	if err != nil {
+		return nil, err
+	}
+
+	data, err := yaml.Marshal(manifest)
+	if err != nil {
+		return nil, err
+	}
+
+	return p.client.CoreV1().Secrets(p.namespace).Create(&corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      secretName,
+			Namespace: p.namespace,
+			Labels: map[string]string{
+				deploymentKeyName: p.deploymentName,
+				versionKeyName:    strconv.Itoa(version),
+				sourceKeyName:     sourceDescription,
+			},
+		},
+		Data: map[string][]byte{
+			manifestKeyName: data,
+		},
+	})
+}
+
+func (p *manifestPersisterImpl) RetrieveVersion(version int) (*corev1.Secret, error) {
+	list, err := p.ListAllVersions()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, secret := range list {
+		ver, err := getVersionFromSecretName(p, secret.Name)
+		if err != nil {
+			return nil, err
+		}
+
+		if ver == version {
+			return &secret, nil
+		}
+	}
+	return nil, fmt.Errorf("unable to find the requested version: %d", version)
+}
+
+func (p *manifestPersisterImpl) RetrieveLatestVersion() (*corev1.Secret, error) {
+	latestVersion, err := getGreatestVersion(p)
+	if err != nil {
+		return nil, err
+	}
+	return p.RetrieveVersion(latestVersion)
+}
+
+func (p *manifestPersisterImpl) ListAllVersions() ([]corev1.Secret, error) {
+	listResp, err := p.client.CoreV1().Secrets(p.namespace).List(metav1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	result := []corev1.Secret{}
+
+	nameRegex := regexp.MustCompile(fmt.Sprintf(`^deployment-%s-\d+$`, p.deploymentName))
+	for _, secret := range listResp.Items {
+		if nameRegex.MatchString(secret.Name) {
+			result = append(result, secret)
+		}
+	}
+	return result, nil
+}
+
+func (p *manifestPersisterImpl) DeleteManifest() error {
+	list, err := p.ListAllVersions()
+	if err != nil {
+		return err
+	}
+
+	for _, secret := range list {
+		err := p.client.CoreV1().Secrets(p.namespace).Delete(secret.Name, &metav1.DeleteOptions{})
+
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (p *manifestPersisterImpl) DecorateManifest(key string, value string) (*corev1.Secret, error) {
+	version, err := getGreatestVersion(p)
+	if err != nil {
+		return nil, err
+	}
+
+	secretName, err := secretName(p, version)
+	if err != nil {
+		return nil, err
+	}
+
+	secret, err := p.client.CoreV1().Secrets(p.namespace).Get(secretName, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	labels := secret.GetLabels()
+	if labels == nil {
+		labels = make(map[string]string)
+	}
+
+	labels[key] = value
+	secret.SetLabels(labels)
+
+	return p.client.CoreV1().Secrets(p.namespace).Update(secret)
+}
+
+func getGreatestVersion(p *manifestPersisterImpl) (int, error) {
```

Comment:
Why not define the private functions on the struct?

Reply:
Thanks for bringing that up. It is actually influenced by a YouTube video I saw on common mistakes in Go. Coming from long time exposure of (I am not ashamed to say it) Java, [this particular part of the video](https://youtu.be/29LLRKIL_TI?t=638) struck home. So I would be interested to hear your preference. Since it does not change any state, it felt like a helper function and therefore was written in that way. I am happy to change it to an unexported method if you want.